### PR TITLE
[HIVEMALL-90] Refine incomplete AUC UDAF implementation

### PR DESCRIPTION
**Pull request by takuti**
## What changes were proposed in this pull request?

Since AUC UDAF (classification) did not work correctly for some specific merge orders, this PR fixes the issue by modifying the UDAF's `merge()` and `terminate()` implementation.

Moreover, unit tests are refined accordingly, and a utility method is created in **HiveUtils**.

## What type of PR is it?

Bug Fix

## What is the Jira issue?

https://issues.apache.org/jira/browse/HIVEMALL-90

## How was this patch tested?

- Unit test
- Manual test on EMR

## How to use this feature?

Nothing has been changed from [current AUC UDAF](https://hivemall.incubator.apache.org/userguide/eval/auc.html).

### CODE REVIEW

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 1]
```diff
@@ -204,21 +256,53 @@ public void merge(AggregationBuffer agg, Object partial) throws HiveException {
...
+            Object tpPrevPartialMapObj = internalMergeOI.getStructFieldData(partial, tpPrevPartialMapField);
+
+            double indexScore = PrimitiveObjectInspectorFactory.writableDoubleObjectInspector.get(indexScoreObj);
+            double area = PrimitiveObjectInspectorFactory.writableDoubleObjectInspector.get(areaObj);
             long fp = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(fpObj);
             long tp = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(tpObj);
             long fpPrev = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(fpPrevObj);
             long tpPrev = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(tpPrevObj);
 
+            Map<Double, Double> areaPartialMap = (Map<Double, Double>) ObjectInspectorFactory.getStandardMapObjectInspector(
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]:
```java
ObjectInspectorFactory.getStandardMapObjectInspector(
  PrimitiveObjectInspectorFactory.writableDoubleObjectInspector,
  PrimitiveObjectInspectorFactory.writableLongObjectInspector
)
```

Invalid casting `Map<DoubleWritable, LongWritable>` to `Map<Double, Double>`

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 2]
```diff
@@ -35,7 +39,9 @@
 import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator;
 import org.apache.hadoop.hive.ql.udf.generic.GenericUDAFEvaluator.AbstractAggregationBuffer;
 import org.apache.hadoop.hive.serde2.io.DoubleWritable;
+import org.apache.hadoop.hive.serde2.lazybinary.LazyBinaryMap;
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: Unused import

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 3]
```diff
@@ -204,21 +256,53 @@ public void merge(AggregationBuffer agg, Object partial) throws HiveException {
...
             long tp = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(tpObj);
             long fpPrev = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(fpPrevObj);
             long tpPrev = PrimitiveObjectInspectorFactory.writableLongObjectInspector.get(tpPrevObj);
 
+            Map<Double, Double> areaPartialMap = (Map<Double, Double>) ObjectInspectorFactory.getStandardMapObjectInspector(
+                PrimitiveObjectInspectorFactory.writableDoubleObjectInspector,
+                PrimitiveObjectInspectorFactory.writableLongObjectInspector).getMap(
+                HiveUtils.castLazyBinaryObject(areaPartialMapObj));
+
+            Map<Double, Long> fpPartialMap = (Map<Double, Long>) ObjectInspectorFactory.getStandardMapObjectInspector(
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: TerminatePartial returns Writable objects but receiving Java objects.

File: core/src/main/java/hivemall/evaluation/AUCUDAF.java [hunk 4]
```diff
@@ -188,13 +234,19 @@ public void iterate(AggregationBuffer agg, Object[] parameters) throws HiveExcep
...
-            partialResult[0] = new DoubleWritable(myAggr.a);
-            partialResult[1] = new DoubleWritable(myAggr.scorePrev);
+            Object[] partialResult = new Object[11];
+            partialResult[0] = new DoubleWritable(myAggr.indexScore);
+            partialResult[1] = new DoubleWritable(myAggr.area);
             partialResult[2] = new LongWritable(myAggr.fp);
             partialResult[3] = new LongWritable(myAggr.tp);
             partialResult[4] = new LongWritable(myAggr.fpPrev);
             partialResult[5] = new LongWritable(myAggr.tpPrev);
+            partialResult[6] = myAggr.areaPartialMap;
```
Comment (myui) [inline core/src/main/java/hivemall/evaluation/AUCUDAF.java]: revise OI. `javaDoubleObjectInspector` instead of `writableDoubleObjectInspector`?

**Comment by myui** [comment]
@takuti terminatePartial/merge OIs are invalid ones.

**Comment by takuti** [comment]
@myui fixed. plz check them.

**Comment by myui** [comment]
@takuti LGTM. Merged to small modifications. Thanks!

**Pull request closed without merging**